### PR TITLE
feat(s3): add s3 ssec option

### PR DESCRIPTION
**Pull request by Yelinz**
add setting to be able to use SSEC when using the S3 storage backend.

S3 will complain if trying to access encrypted files without a key or unencrypted with a key.

Therefore enabling encryption has to be coupled with a reupload of all files with the correct setting. (management command will do it)

### CODE REVIEW

File: document_merge_service/api/management/commands/dms_encrypt_templates.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                    "Encryption is not enabled. Skipping encryption of templates."
+                )
+            )
+
+        failed_templates = []
+
+        # flip between default and encrypted storage to have the correct parameters in the requests
+        encrypted_storage = storages.create_storage(settings.STORAGES["default"])
+        normal_storage_setting = settings.STORAGES["default"]
+        del normal_storage_setting["OPTIONS"]["object_parameters"]
```
Comment (fugal-dy) [inline document_merge_service/api/management/commands/dms_encrypt_templates.py:25]: That will throw a key error, if `OPTIONS` or `object_parameters` don't exist.

Either only `del` if the item exists or raise `ConfigurationError` if that's required.

File: document_merge_service/api/management/commands/dms_encrypt_templates.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                self.style.WARNING(
+                    "Encryption is not enabled. Skipping encryption of templates."
+                )
+            )
+
+        failed_templates = []
+
+        # flip between default and encrypted storage to have the correct parameters in the requests
+        encrypted_storage = storages.create_storage(settings.STORAGES["default"])
+        normal_storage_setting = settings.STORAGES["default"]
```
Comment (fugal-dy) [inline document_merge_service/api/management/commands/dms_encrypt_templates.py:24]: I'd prefer to name `normal` something more desciptive like `unencrypted` or `plain` or something.

File: document_merge_service/api/tests/test_encrypt_templates.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+def test_encrypt_templates(db, settings, mocker, template_factory):
+    template_factory(template=django_file("docx-template.docx"))
+
+    settings.DMS_ENABLE_AT_REST_ENCRYPTION = True
+    settings.STORAGES["default"]["BACKEND"] = "storages.backends.s3.S3Storage"
+    settings.STORAGES["default"]["OPTIONS"] = settings.S3_STORAGE_OPTIONS
+    settings.STORAGES["default"]["OPTIONS"]["object_parameters"] = {
+        "SSECustomerKey": "x" * 32,
+        "SSECustomerAlgorithm": "AES256",
+    }
```
Comment (fugal-dy) [inline document_merge_service/api/tests/test_encrypt_templates.py:19]: Those nested settings will probably leak to other tests at some point.

You should either monkeypatch those or make copy that will be lost after the test.

File: document_merge_service/api/tests/test_encrypt_templates.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+def test_encrypt_templates_failed(db, settings, mocker, template_factory):
+    template_factory(template=django_file("docx-template.docx"))
+
+    settings.DMS_ENABLE_AT_REST_ENCRYPTION = True
+    settings.STORAGES["default"]["BACKEND"] = "storages.backends.s3.S3Storage"
+    settings.STORAGES["default"]["OPTIONS"] = settings.S3_STORAGE_OPTIONS
+    settings.STORAGES["default"]["OPTIONS"]["object_parameters"] = {
+        "SSECustomerKey": "x" * 32,
+        "SSECustomerAlgorithm": "AES256",
+    }
```
Comment (fugal-dy) [inline document_merge_service/api/tests/test_encrypt_templates.py:52]: Those are leaky, too.

**Review by fugal-dy — COMMENTED** [review]
I had a closer look on the command and suggest some safeguards. On to I found some potentially leaky test settings.

File: document_merge_service/api/management/commands/dms_encrypt_templates.py (hunk 1 above)
Comment (Yelinz) [reply document_merge_service/api/management/commands/dms_encrypt_templates.py]: Added check

File: document_merge_service/api/management/commands/dms_encrypt_templates.py (hunk 2 above)
Comment (Yelinz) [reply document_merge_service/api/management/commands/dms_encrypt_templates.py]: Done

File: document_merge_service/api/tests/test_encrypt_templates.py (hunk 3 above)
Comment (Yelinz) [reply document_merge_service/api/tests/test_encrypt_templates.py]: changed to copy

File: document_merge_service/api/tests/test_encrypt_templates.py (hunk 4 above)
Comment (Yelinz) [reply document_merge_service/api/tests/test_encrypt_templates.py]: changed to copy

File: document_merge_service/api/management/commands/dms_encrypt_templates.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            "OPTIONS" not in unencrypted_storage_setting
+            or "object_parameters" not in unencrypted_storage_setting["OPTIONS"]
+        ):
+            raise ImproperlyConfigured(
+                "Encryption is enabled but no object_parameters found in the storage settings."
+            )
+        del unencrypted_storage_setting["OPTIONS"]["object_parameters"]
+        unencrypted_storage = storages.create_storage(unencrypted_storage_setting)
+
+        query = Template.objects.all()
```
Comment (fugal-dy) [inline document_merge_service/api/management/commands/dms_encrypt_templates.py:36]: How do you know you're not encrypting an already encrypted file/template?

Comment (Yelinz) [reply document_merge_service/api/management/commands/dms_encrypt_templates.py:36]: There is no way, you just have to assume everything is plain text or encrypted.

If its already encrypted the fetching wont work, as we dont send the ssec key.

**Comment by czosel** [comment]
@fugal-dy thanks for the review! I took the liberty to resolve the points after checking that they were adressed (we'd like to ship this :wink: )

**Review by czosel — APPROVED** [review]

**Pull request merged**